File: app/src/main/java/br/com/vicentec12/mygames/ui/console/ConsoleViewModel.java
package br.com.vicentec12.mygames.ui.console;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.List;

import javax.inject.Inject;

import br.com.vicentec12.mygames.data.model.ConsoleWithGames;
import br.com.vicentec12.mygames.data.source.console.ConsoleDataSource;

public class ConsoleViewModel extends ViewModel {

    private final int CHILD_PROGRESS = 0;
    private final int CHILD_RECYCLER = 1;
    private final int CHILD_TEXT = 2;

    private final MutableLiveData<Integer> _message = new MutableLiveData<>();
    private final MutableLiveData<Integer> _viewFlipperChild = new MutableLiveData<>();
    private final MutableLiveData<List<ConsoleWithGames>> _consoles = new MutableLiveData<>();

    private final ConsoleDataSource mConsoleRepository;

    @Inject
    public ConsoleViewModel(ConsoleDataSource mConsoleRepository) {
        this.mConsoleRepository = mConsoleRepository;
    }

    public void listConsoles() {
        _viewFlipperChild.setValue(CHILD_PROGRESS);
        mConsoleRepository.listConsolesWithGames(new ConsoleDataSource.OnConsolesWithGamesListedCallback() {
            @Override
            public void onSucess(int message, List<ConsoleWithGames> consolesWithGames) {
                _consoles.setValue(consolesWithGames);
                _viewFlipperChild.setValue(CHILD_RECYCLER);
            }

            @Override
            public void onErro(int message) {
                _message.setValue(message);
                _viewFlipperChild.setValue(CHILD_TEXT);
            }
        });
    }

    public MutableLiveData<List<ConsoleWithGames>> getConsoles() {
        return _consoles;
    }

    public MutableLiveData<Integer> getViewFlipperChild() {
        return _viewFlipperChild;
    }

    public MutableLiveData<Integer> getMessage() {
        return _message;
    }

}

File: app/src/main/java/br/com/vicentec12/mygames/ui/console/ConsoleViewModel.kt
package br.com.vicentec12.mygames.ui.console

import androidx.annotation.StringRes
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.com.vicentec12.mygames.domain.model.Console
import br.com.vicentec12.mygames.domain.use_case.console.ListWithGamesUseCase
import br.com.vicentec12.mygames.extensions.error
import br.com.vicentec12.mygames.extensions.success
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ConsoleViewModel @Inject constructor(
    private val mListWithGamesUseCase: ListWithGamesUseCase
) : ViewModel() {

    private val _uiState = MutableLiveData<UiState>(UiState.Loading)
    val uiState: LiveData<UiState> = _uiState

    fun listConsoles() {
        viewModelScope.launch {
            _uiState.value = UiState.Loading
            mListWithGamesUseCase().error { mResult ->
                _uiState.value = UiState.Error(mResult.message)
            }.success { result ->
                _uiState.value = UiState.Consoles(result.data.orEmpty(), result.message)
            }
        }
    }

    companion object {

        private const val CHILD_PROGRESS = 0
        private const val CHILD_CONSOLES = 1
        private const val CHILD_MESSAGE = 2

    }

    sealed class UiState {
        object Loading : UiState()
        class Consoles(val mConsoles: List<Console>, @StringRes val mMessage: Int) : UiState()
        class Error(@StringRes val mMessage: Int) : UiState()
    }

}